### ISSUE TITLE
[FEAT] - Separately negotiated version number for NTC queries

### ISSUE THREAD

**Issue by nfrisby** [IntersectMBO/ouroboros-consensus#866]
Imagine some innovative small company wants to use only the _abstract_ parts of ouroboros-network, ouroboros-consensus, and cardano-ledger. They want to write their own blocks, with their own (Ouroboros-compatible) ledger rules and queries, entirely separate from Byron, Shelley, Allegra, etc.

Moreover, suppose they're not going to maintain forks of our code. They'll use whatever the latest version of Ouroboros is via our packages, and build everything else around those dependencies.

Today's `NodeToNodeVersion` and `NodeToClientVersion` data types in ouroboros-network prevent this use case from being sustainable, because these hypothetical devs (who by assumption are using the Diffusion Layer) would be unable to introduce new negotiable handshake versions. (They'd have to wait until Cardano does in order to ever release a feature that requires a bump to the negotiated version🤦.)

Proposal:

> We need a second, separately negotiated version number that is not at all restricted by the core Ouroboros code (ie just a natural number). Its semantics is determined solely by the block type.

- The two existing data types are irreplaceable for versioning the block-agnostic core Ouroboros behaviors: the mini protocols themselves, eg their codecs, eg block diffusion pipelining (in ChainSync, NTN), eg whether `LocalTxMonitor` is enabled (NTC), etc.
- But they should not affect any block-specific behaviors.
- This new version number would govern at least the evolution of blocks' Node-to-Client queries.
- I currently suspect the new version number would not be useful for any block-specific aspects of Node-to-Node communication, since I anticipate those are all fully handled by the on-chain data, eg the chain's protocol version.

**Comment by nfrisby** [comment]
@amesgen @coot @njd42 @dcoutts @abailly @KtorZ What do you think? Am I overlooking some part of the handshake that already supports this (idealized) use case?

**Comment by nfrisby** [comment]
Perhaps this should happen as an initialization phase of `LocalStateQuery` itself?

**Comment by abailly-iohk** [comment]
I am fully supportive of the use case you envision, eg. making alternative implementations of Cardano network participants first class citizen if I understood correctly. I am not sure why you would need a different number, though. What I would do here is to standardise the protocol separately from the implementation, use whatever versioning scheme makes sense and from that point on allow a separate evolution of the protocol specification and the implementations.
This could apply equally well to N2N and N2C protocols, I think, as some nodes could decide to support exchanging some data which other nodes do not support yet (eg. Mithril signatures, Peras votes) and the network could evolve in a more incremental way.
Perhaps this vision is very naive :)

**Comment by coot** [comment]
I also agree that it's a really good idea to support alternative use cases.

If this is really an alternative blockchain they ought to use alternative `NodeTo{Node,Client}Version` versions and we could generalise our protocol to be parametrised over the version types.  This would allow one to govern their case by themselves, but would require them to modify `cardano-{node,cli}`, which they'd need to do anyway.

`KeepAlive` is fundamental for at least the first of the two reasons:

* keep the network running,
* do measurements which are used by `BlockFetch`.

**Comment by nfrisby** [comment]
@abailly said

> making alternative implementations of Cardano network participants first class citizen if I understood correctly

but that's actually not what I meant. I meant using the Ouroboros code to run a different non-Cardano blockchain.

I suspect that misunderstanding supersedes the rest of your comment, unfortunately?

**Comment by abailly-iohk** [comment]
> @abailly said
> but that's actually not what I meant. I meant using the Ouroboros code to run a different non-Cardano blockchain.

😢 
 
> I suspect that misunderstanding supersedes the rest of your comment, unfortunately?

😂

**Comment by abailly-iohk** [comment]
Please ignore my comment then @nfrisby

**Comment by nfrisby** [comment]
@coot said

> they ought to use alternative `NodeTo{Node,Client}Version` versions and we could generalise our protocol to be parametrised over the version types

I'll elaborate what I think you're proposing. Please let me know if I'm wrong. (I'm not sure I agree with your proposal :), but I want to first confirm I get what you're saying.)

> The hypothetical devs' (using our Ouroboros code to run their own non-Cardano chain) would have to provide types `MyVersionNT{N,C}`, a function `MyVersionNTN -> DiffusionLayer.NodeToNodeVersion`, and a function `MyVersionNTC -> DiffusionLayer.NodeToNodeClient` when invoking the Diffusion/Consensus layers. And we would have adjusted the Diffusion/Consensus code to additionally pipe the `MyVersionNTC` version into their queries. The Handshake protocol would negotiate `MyVersionNTN` and `MyVersionNTC` instead of negotiating `NodeTo{Node,Client}Version`.

**Comment by nfrisby** [comment]
I had a nice chat with @njd42 and @karknu. I think I can better present my proposal now. In particular, they had two pieces of clarifying feedback:

- One handshake to rule them all. The Diffusion Layer has had the design goal that the version negotiated via Handshake should determine everything the node needs to know about the peer.

- Keep it simple. We'd rather not introduce these abstractions until we have a customer/user/community who actually benefits from them.

(My current proposal would only weaken that design goal to "everything the Diffusion Layer needs to know about the peer". And the scope of my proposal is: "this is how we should think about versioning NTC queries", but it's totally reasonable  continue conflating their version with the Handshake version until we actually have a worthwhile reason to introduce the proposed degree of freedom.

In light of that very helpful feedback, I'll update my proposal's text.

**Comment by nfrisby** [comment]
I'm now rewriting the Issue description, so I'm archiving the original one (to which all the above replies were written) in this comment.

---

Imagine some innovative small company wants to use only the _abstract_ parts of ouroboros-network, ouroboros-consensus, and cardano-ledger. They want to write their own blocks, with their own (Ouroboros-compatible) ledger rules and queries, entirely separate from Byron, Shelley, Allegra, etc.

Moreover, suppose they're not going to maintain forks of our code. They'll use whatever the latest version of Ouroboros is via our packages, and build everything else around those dependencies.

Today's `NodeToNodeVersion` and `NodeToClientVersion` data types in ouroboros-network prevent this use case from being sustainable, because these hypothetical devs (who by assumption are using the Diffusion Layer) would be unable to introduce new negotiable handshake versions. (They'd have to wait until Cardano does in order to ever release a feature that requires a bump to the negotiated version🤦.)

Proposal:

> We need a second, separately negotiated version number that is not at all restricted by the core Ouroboros code (ie just a natural number). Its semantics is determined solely by the block type.

- The two existing data types are irreplaceable for versioning the block-agnostic core Ouroboros behaviors: the mini protocols themselves, eg their codecs, eg block diffusion pipelining (in ChainSync, NTN), eg whether `LocalTxMonitor` is enabled (NTC), etc.
- But they should not affect any block-specific behaviors.
- This new version number would govern at least the evolution of blocks' Node-to-Client queries.
- I currently suspect the new version number would not be useful for any block-specific aspects of Node-to-Node communication, since I anticipate those are all fully handled by the on-chain data, eg the chain's protocol version.

**Comment by nfrisby** [comment]
I have now updated the Issue description with a much more fleshed out (and longer :/) motivation/proposal. cc @coot @njd42 @karknu
